### ISSUE TITLE
Glitching in WASAPI in shared mode

### ISSUE THREAD

**Issue by sveinse** [PortAudio/portaudio#303]
I'm experiencing significant glitching when using WASAPI from master in shared mode. If using exclusive mode or a full-duplex audio stream, the audio doesn't glitch. 

I'm using PA from Python using the interface library [pyaudio](https://github.com/sveinse/pyaudio). I'm not excluding performance issues from the python environment, but all the other hostAPIs works smoothly without glitching.

The glitching is typically manifested as bursts of 4096 samples played back successfully and then 220-230 samples of pause in between. The pause interval varies. PA is called with `framesPerBuffer` of 1024. It improves the occurrence of glitching by increasing this number, but it doesn't remove the glitching altogether, even with very large values such as 65536.

For reference my environment is:

* Windows 10, 2004 version
* Compiled with MS VS 2019 Community Edition
* Applied patch for building (for MSVC 2019 and to removed ASIO): https://gist.github.com/sveinse/7a3442d6f8444b95c4084a7172ec5fdb
* Compile command (from git bash):

      '/c/Program Files (x86)/Microsoft Visual Studio/2019/Community/MSBuild/Current/Bin/MSBuild.exe' \ 
          build\\msvc\\portaudio.sln \
          //property:GenerateFullPaths=true \
          //t:build \
          //consoleloggerparameters:NoSummary \
          //p:Configuration=Debug \
          //p:Platform=win32

**Comment by sveinse** [comment]
Some observations:

* PortAudio from `master` glitches on WASAPI playback
* PortAudio from `pa_stable_v190600_20161030` does not glitch
* WASAPI used in exclusive mode does not glitch
* WASAPI used in full duplex mode does not glitch

I made a git bisect to determine which commit caused the gliching behavior in WASAPI. The breaking commit commited on Sep 6 2018 is:

c3fee03ff143feca50dcbc53b89670f251bbce24

**Comment by sveinse** [comment]
I think I have found evidence that this is related to sleep timing setup in `ProcThreadPoll()` in `pa_win_wasapi.c` and that the timing amounts have doubled in master causing these glitches.

[pa_win_wasapi.c](https://github.com/PortAudio/portaudio/blob/master/src/hostapi/wasapi/pa_win_wasapi.c) implements [`GetFramesSleepTime()`](https://github.com/PortAudio/portaudio/blob/master/src/hostapi/wasapi/pa_win_wasapi.c#L916-L949) to calculate thread sleep time. This function is used in the main worker thread to calculate the sleep time between each iteration in poll mode.

https://github.com/PortAudio/portaudio/blob/master/src/hostapi/wasapi/pa_win_wasapi.c#L5764-L5802

The change that cause WASAPI to start glitching is that the sleep times returned from `GetFramesSleepTime()` has been doubled. This works without glitching: 

https://github.com/PortAudio/portaudio/blob/d7a4cb482dc7470c1863bd0ff973ca2c6aefb70c/src/hostapi/wasapi/pa_win_wasapi.c#L855 

While this does not work and produces glitching:

https://github.com/PortAudio/portaudio/blob/c3fee03ff143feca50dcbc53b89670f251bbce24/src/hostapi/wasapi/pa_win_wasapi.c#L896

My test setup runs at 48kHz, with `framesPerBuffer` at 1024 (21.3ms). (WASAPI prints `framesPerUser[ 1024 ] framesPerHost[ 2048 ] latency[ 42.67ms ]`). This is a setup of the observed sleep values (`sleep_ms`) used in `ProcThreadPoll()` under different usage conditions:

|  What | Master | Stable |
| --- | --- | --- |
| Playback | 21ms (glitch) | 10ms (OK) |
| Recording | 2ms (OK) | 2ms (OK) |
| Full-duplex | 2ms (OK) | 2ms (OK) |
| Exclusive mode playback | 21ms (OK) | 10ms (OK) |

Observations:
* Recording is clipped to 2 ms sleep by intentional design which is the reason why no glitching is obseved on record
* In full-duplex the sleep is selects the lowest sleep value, hence because of the previous item, no glitching is observed in full-duplex mode
* Exclusive mode seems more resilient against glitching, as the sleep time is the same as shared mode that glitches.

**Comment by sveinse** [comment]
I have implemented this workaround in my own portaudio build which I use from my Python application and it works fine:

https://github.com/sveinse/portaudio/commit/4a88ee7813d007c779defe268e42a3b302518d9d

In essence it reverts the thread sleep delay to the same values as in the stable version of PA.

**Comment by dmitrykos** [comment]
@sveinse , thank you for raising this issue. In my setup I do not experience any glitching but may be it happens when latency is the minimal for the Shared Mode, e.g. as in your case - 21ms (glitch).

Your workaround for `ProcThreadPoll` looks fine to me as similar approach is already used for the blocking API (`ReadStream`, `WriteStream`). Could you submit the merge request to incorporate this change?

**Comment by sveinse** [comment]
@dmitrykos I created a PR for this fix.

I'm not sure what makes one system (like yours) not glitch and others is glitching a lot. With the old solution it does not matter what the `framesPerBuffer` value is, it always glitches, not only on low-values.

Yet, I'm not completely convinced that this fix is complete. I ran a series of tests after the PR and I still experience gliching in some ranges of `framesPerBuffer`:

| `framesPerUser` | `framesPerHost` | latency | glitches? | |
| --- | --- | --- | --- | --- |
| 4096 | 8192 | 171ms | No |
| 2048 | 4096 | 85ms | No |
| 1024 | 2048 | 43ms | No |
| 512 | 1056 | 22ms | Yes |
| 256 | 1056 | 22ms | Yes | 
| 128 | 1056 | 22ms | No | (usec sleep) |
| 64 | 1056 | 22ms | No | (usec sleep) |
| 6 | 1056 | 22ms | No | (usec sleep) |

In my system any `framesPerBuffer` value between 129 - 799 glitches, regardless of which audio device used. It starts working seamless again below 128 and notice that it keeps track with seamless playback the whole way down to 6 samples per iteration (and this includes jumps into Python runtime code).

**Comment by dmitrykos** [comment]
@sveinse I merged the fix as a quick solution for now. Although it fixes the problem you observed, we still need to find the reason of the underruns/glitches.

The line:
```c
sleep_ms_out = GetFramesSleepTime(stream->out.framesPerBuffer, stream->out.wavex.Format.nSamplesPerSec);
````
will cause sleeps for a duration of the user buffer, e.g. we will always sleep less than the duration of the host (WASAPI's side) buffer which size is reflected in `framesPerHostCallback` variable. According to the official docs the code there waits half of the host buffer duration (https://docs.microsoft.com/en-us/windows/win32/coreaudio/rendering-a-stream):
```cpp
// Sleep for half the buffer duration.
Sleep((DWORD)(hnsActualDuration/REFTIMES_PER_MILLISEC/2));
````

so our implementation (before this fix) had to be stable in terms of absence of the underruns.

In your failed tests with this fix we will sleep for:
- 512: 5 ms = 512 / 48000 / 2
- 256: 2 ms = 256 / 48000 / 2

and it does not provide any clue why glitches still exist because the host's buffer duration is 22 ms, so we have plenty of time. Could it be due to a jitter of the system timer/Sleep?

**Comment by dmitrykos** [comment]
@sveinse, would you please make an experiment and replace line 3859 to such:
```c
bufferMode = paUtilBoundedHostBufferSize;
```
and test with `framesPerUser` with 512 and 256.

Let's check if we do not have any problem inside the rendering loop, specifically this area starting at line 5934:
```c
if (stream->bufferMode == paUtilFixedHostBufferSize)
{
    while (frames >= stream->out.framesPerBuffer)
    {
        if ((hr = ProcessOutputBuffer(stream, processor, stream->out.framesPerBuffer)) != S_OK)
        {
            LogHostError(hr);
            goto thread_error;
        }

        frames -= stream->out.framesPerBuffer;
    }
}
```

Here I got worried that if this condition `while (frames >= stream->out.framesPerBuffer)` is false right after the `_PollGetOutputFramesAvailable` then we can run into the underrun due to sleeping again.

**Comment by sveinse** [comment]
@dmitrykos setting `bufferMode` improves the stability a lot, but it still glitches. It goes from a constant stream of glitches, to a more randomly occurring instances, up to 10 seconds apart. It also seems stability (glitching) occurs at the lower `framesPerBuffer` values with this new.

BTW I'm monitoring and printing if any over- or underrun flags are reported to the user callback function, and they are not flagged in any of the glitches. So either WASAPI is oblivious to the underruns or its missing an implementation of reporting this to the user callback function. I haven't checked what it does.

**Comment by dmitrykos** [comment]
@sveinse, thank you for a test! 

Even if stability is improved but there are glitches left then `paUtilFixedHostBufferSize` rendering loop seems to be ok but it is causing more instability due to requesting host buffer in a fixed number of chunks (`stream->out.framesPerBuffer`).

What if you make another experiment and hardcode `sleep_ms_out` to be equal 11, e.g. it will be exactly 50% of the host buffer that is: 1056 / 48000 / 2 = 11. Earlier you reported that your test with `framesPerUser == 1024` was without glitching and the `sleep_ms_out` in that case was 10. So actually could you check 10 an 11 ms and with `bufferMode` reverted to `paUtilFixedHostBufferSize`.

**Comment by sveinse** [comment]
@dmitrykos 

Given the apparent host native size of 1056, I tried to hardcode `sleep_ms` as suggested. If I use `framesPerBuffer` of 528, I find no values of `sleep_ms` that results in stable playback (tried 11, 10, 8, 6, 5, 4). Same thing with 264: no stable playback. At 132 it is glitch free at value 1, but not at 2. At 66 something strange happens: Then I get stable playback with any sleep values up to approx 10!

In my post above, the table seems to hint that usec sleep is what resolves the glitching when `framesPerBuffer` gets low. So I tested that and it does not. It does glitch even with some values where usec sleep is selected. So that can be crossed off the list.

**Comment by sveinse** [comment]
@dmitrykos 

As for the `while (frames >= stream->out.framesPerBuffer)` I did a crude test:

```c
if (stream->bufferMode == paUtilFixedHostBufferSize)
{
	UINT32 count = 0;
	while (frames >= stream->out.framesPerBuffer)
	{
		count++;
		if ((hr = ProcessOutputBuffer(stream, processor, stream->out.framesPerBuffer)) != S_OK)
		{
			LogHostError(hr);
			goto thread_error;
		}

		frames -= stream->out.framesPerBuffer;
	}
	printf("%d,", count);
}
```

(Don't flog me for the printf() within real-time-ish code, but it does work sufficiently to highlight the results :D)

```
WASAPI::OpenStream(output): framesPerUser[ 528 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
2,0,2,0,2,0,2,0,2,1,1,2,0,2,0,2,0,2,1,1,2,0,2,0,2,0,2,1,1,2,0,2,0,2,1,1,2,0,2,0,2,0,2,1,1,2,0,2,1,1,2,0,2,1,1,2,0,2,0,2,1,1,2,0,2,1,1,2,0,2,0,2,1,1,2,0,2,0,2,
```

So it definitely hits the while loop without enough data available, resulting in no callbacks to `ProcessOutputBuffer()`. In another run I looked at `frames` from `_PollGetOutputFramesAvailable` and I get highly variable numbers:

```
1056 960 912 1056 480 1056 480 1056 960 912 1056 480 1056 480 1056 960 912 1056 480 1056 480 1056 960 912 1056 480 1056 960 912 1056 480 1056 960 912 1056 480 1056 480 1056 480 1056 960 912 1056 480 1056 960 912 1056 480 1056 480 1056 960 912 1056 480 1056 960 912 1056 480 1056 960 912 1056 480 1056 960 912 1056 480 1056 480 1056 960 912
```

**Comment by kmatheussen** [comment]
> BTW I'm monitoring and printing if any over- or underrun flags are reported to the user callback function, and they are not flagged in any of the glitches. So either WASAPI is oblivious to the underruns or its missing an implementation of reporting this to the user callback function. I haven't checked what it does.

Thank you both for your work. Non-reported glitches is my biggest concern about using the current WASAPI driver in portaudio.

**Comment by kmatheussen** [comment]
I've looked a little bit at the code now, and I'm probably wrong, or maybe not, but as far as I can see, the audio processing itself (called in the function "WaspiHostProcessingLoop") happens in the same thread as the one that sleeps (the function "ProcThreadPoll"). Doesn't this mean that when you spend half the time sleeping you also can't use more than 50% CPU to process audio?

**Comment by dmitrykos** [comment]
@sveinse thank you for your experiment. To my view we can have underruns happening here:

`2,0,2,0`

e.g.  after having no frames to process on the next cycle we get 2 buffers to fill (1056 / 528 = 2) which are equal to a full host buffer and it is a problem, I would rather expect sequence `1,0,1,0` or `1,0,1,1` that would mean that we usually get 1/2 of the host buffer. Here I assumed that in your test `stream->out.framesPerBuffer` equals 528.

I did some more experiments and tried to make busy waiting once we get 0 like this:
```c
UINT32 next_sleep = sleep_ms;
UINT32 repeat = 0;

// Processing Loop
while (WaitForSingleObject(stream->hCloseRequest, next_sleep) == WAIT_TIMEOUT)
{
    // Get next sleep time
    if (sleep_ms == 0)
        next_sleep = ThreadIdleScheduler_NextSleep(&scheduler);

    for (i = 0; i < S_COUNT; ++i)
    {
        // Process S_INPUT/S_OUTPUT
        switch (i)
        {
        // Input stream
        case S_INPUT: {

            if (stream->captureClient == NULL)
                break;

            if ((hr = ProcessInputBuffer(stream, processor)) != S_OK)
            {
                LogHostError(hr);
                goto thread_error;
            }

            break; }

        // Output stream
        case S_OUTPUT: {

            UINT32 frames;
            if (stream->renderClient == NULL)
                break;

            // Get available frames
            if ((hr = _PollGetOutputFramesAvailable(stream, &frames)) != S_OK)
            {
                LogHostError(hr);
                goto thread_error;
            }

            // Output data to the user callback
            if (stream->bufferMode == paUtilFixedHostBufferSize)
            {
                UINT32 count = 0;
                while (frames >= stream->out.framesPerBuffer)
                {
                    count++;
                    if ((hr = ProcessOutputBuffer(stream, processor, stream->out.framesPerBuffer)) != S_OK)
                    {
                        LogHostError(hr);
                        goto thread_error;
                    }

                    frames -= stream->out.framesPerBuffer;
                }

                if (count == 0)
                {
                    repeat++;
                    next_sleep = 0;
                    continue;
                }
                else
                {
                    //printf("%d,", repeat);
                    repeat = 0;
                }
            }
            else
            if (frames != 0)
            {
                if ((hr = ProcessOutputBuffer(stream, processor, frames)) != S_OK)
                {
                    LogHostError(hr);
                    goto thread_error;
                }
            }

            break; }
        }
    }

    next_sleep = sleep_ms;
}
```

Explanation:

Once we get 0 we should not sleep again (may be it was a corner case and buffer is already ready) as it can result in the underrun and would rather do a short busy waiting by clearing `next_sleep` to 0 temporarily. I recorded how many repeated cycles the busy loop is doing and it is 1-2 cycles that is pretty ok for a such configuration:

```c
framesPerUser[ 300 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]

// 1,2,2,2,2,2,2,2,1,2,1,2,1,1,2
```

Could you please replace your implementation with this piece of code and check if it improves anything on your machine.

**Comment by dmitrykos** [comment]
@kmatheussen, your assumption is correct but we need to get some space from WASAPI's host buffer to write into it and thus code must sleep/wait for some piece of the host buffer to become available for writing into it. 

To solve the problem you are mentioning there should be a 2-nd audio rendering thread that is doing audio processing on the app's side, it would write to own temporary buffer/-s and once WASAPI's host buffer is available the PA's callback function would simply copy ready audio samples prepared by the 2-nd audio processing thread. 

With such approach you can utilize CPU effectively but it is outside the PA's scope and rather a design of the application which is using PA as audio API.

**Comment by sveinse** [comment]
@kmatheussen, @dmitrykos , with respect to amount of wall-time for user processing/callback of audio: How is this implemented for other host API? Do they also sleep equal amounts? I guess, from a principle point of view, this is the price of polling. One have to wait, check, wait and so on. You will lose granularity on this. The other alternative is to wake up the thread when there is data available -- which is what the event-based variant is all about, right?

Could the sleep be dynamic? In principle the outer loop is aware of when data should be available (being more or less monotonic with respect to wall-time). Could the sleep after user callback be dynamic in the sense that it sleeps the remainder of the timeslot rather than a fixed pre-calculated amount?

**Comment by dmitrykos** [comment]
> Could the sleep be dynamic?

I think so. We could calculate how much time does `ProcessOutputBuffer` take and substract it from the `next_sleep`.

**Comment by dmitrykos** [comment]
@sveinse I quickly made the change you proposed that substructs time spent in the processing from the `next_sleep`, could you please check if anything changed in your setup with this piece of code?:

```c
INT32 next_sleep = sleep_ms;
UINT32 repeat = 0;

// Processing Loop
while (WaitForSingleObject(stream->hCloseRequest, next_sleep) == WAIT_TIMEOUT)
{
    double start_time = PaUtil_GetTime();

    for (i = 0; i < S_COUNT; ++i)
    {
        // Process S_INPUT/S_OUTPUT
        switch (i)
        {
        // Input stream
        case S_INPUT: {

            if (stream->captureClient == NULL)
                break;

            if ((hr = ProcessInputBuffer(stream, processor)) != S_OK)
            {
                LogHostError(hr);
                goto thread_error;
            }

            break; }

        // Output stream
        case S_OUTPUT: {

            UINT32 frames;
            if (stream->renderClient == NULL)
                break;

            // Get available frames
            if ((hr = _PollGetOutputFramesAvailable(stream, &frames)) != S_OK)
            {
                LogHostError(hr);
                goto thread_error;
            }

            // Output data to the user callback
            if (stream->bufferMode == paUtilFixedHostBufferSize)
            {
                UINT32 count = 0;
                while (frames >= stream->out.framesPerBuffer)
                {
                    count++;
                    if ((hr = ProcessOutputBuffer(stream, processor, stream->out.framesPerBuffer)) != S_OK)
                    {
                        LogHostError(hr);
                        goto thread_error;
                    }

                    frames -= stream->out.framesPerBuffer;
                }

                if (count == 0)
                {
                    repeat++;
                    next_sleep = 0;
                    continue;
                }
                else
                {
                    //printf("%d,", repeat);
                    repeat = 0;
                }
            }
            else
            if (frames != 0)
            {
                if ((hr = ProcessOutputBuffer(stream, processor, frames)) != S_OK)
                {
                    LogHostError(hr);
                    goto thread_error;
                }
            }

            break; }
        }
    }

    // Get next sleep time
    if (sleep_ms == 0)
        next_sleep = ThreadIdleScheduler_NextSleep(&scheduler);
    else
        next_sleep = sleep_ms;

    // Update next sleep time dynamically depending on how much time we spent in processing
    next_sleep -= (INT32)((PaUtil_GetTime() - start_time) * 1000);
    if (next_sleep < 0)
        next_sleep = 0;
}
```

**Comment by sveinse** [comment]
@dmitrykos I quickly tested them both (with and without dynamic sleep), and none of them resolves the glitching issues, unfortunately.

I'm curious if its my HW environment that make this behave differently from your, or if its related to the type of callback system I use (i.e. python). It would be really great if you could download the standalone executable https://github.com/sveinse/elns-release/releases/download/v1.1-b3/elns-1.1-b3-standalone.exe and test if WASAPI glitches in your system. Select "WASAPI" in audio system, find a suitable wav or flac file as input and select an appropriate output audio device and then press play. (I hope you're not set off by my executable, its an application I author and maintain. Is a py application including bundled portaudio binaries.)

**Comment by dmitrykos** [comment]
@sveinse I tested your app and can hear glitches on all latencies except 128. I used 1 kHz sine in WAV as it is very easy to hear glitches.

Nevertheless the work that was done above is anyway useless and I tend to include it into the actual implementation:

1. Do short busy waiting if no frames available yet while we expected them;
2. Sleep dynamically depending on spent time for processing.

In my tests I noticed that with the fix you proposed `sleep_ms /= 2` the busy waiting (1.) is more often (expected) because we sleep 2 times shorter time than `stream->out.framesPerBuffer` and to my view it needs to be removed in favor of 1. and 2. changes.

**Comment by sveinse** [comment]
@dmitrykos I made a brute force test by setting `while (WaitForSingleObject(stream->hCloseRequest, 0) == WAIT_TIMEOUT)`, which essentially makes the whole thing a large busy loop. And no surprise really, it works with all `framesPerBuffer` sizes. Interestingly it is capable of having `framesPerBuffer` as low as 2 samples (@48kHz), which each will jump into python runtime execution, do some numpy operations and return. It does spend considerable CPU, but it reveals the true computing power of recent CPUs.

This tells me this seems to be purely timing related. `WaitForSingleObject` is possibly either waiting too long or jitters too much. Is the granularity of ms resolution too coarse? E.g. when the frame size dictates e.g. 5.5ms, not 5 or 6 ms. - But no, it still fails when being switched over to the usec granularity.

Anyways, I hoped this could be useful input.

**Comment by kmatheussen** [comment]
(continuing disussion from #305 here)
@dmitrykos Of course increased CPU consumption can be a problem. My question is if sleeping 2ms makes any difference to care about? No one cares if the portaudio driver uses 0.0001% more CPU, for instance. I'm suggestion that we measure this, and use 2ms or even 1ms if it doesn't make any notable difference.

**Comment by kmatheussen** [comment]
@dmitrykos You wrote: "you could do audio rendering not inside the PA's thread but in some other thread which would deliver ready buffers/data into the PA's callback thread"

Are you serious? Of course you have to do processing in a separate thread if your own code doesn't behave in a proper RT manner, but you should definitely not have to do it because PA doesn't behave in a proper RT manner.

**Comment by kmatheussen** [comment]
Regarding the suggestion for dynamic timing above, I think the problem might be that you are too optimistic about slack. There is no slack there. Some time will be spent in measuring time, waiting, etc., and then I think you will lose the deadline. I think it should work if you do something this instead:

```
next_sleep -= (INT32)((PaUtil_GetTime() - start_time) * 1000);
INT32 half = next_sleep / 2;
if (half <= 2)
   next_sleep = 0;
else
   next_sleep = half;
```

**Comment by dmitrykos** [comment]
@sveinse ok, then probably there is something specific about timing. Is it possible that somewhere in your app `timeBeginPeriod` is called with some >1 value? You could also try adding a call `timeBeginPeriod(1);` inside the `ProcThreadPoll` to double check if system timer resolution wasn't changed ([here ](http://www.windowstimestamp.com/description)is a good overview of the available time API).

Then for another experiment you could add `Sleep(next_time);` inside the `while (WaitForSingleObject(stream->hCloseRequest, 0) == WAIT_TIMEOUT)` loop just to check what happens and whether glitches return or not. If they return another experiment would be to measure how many milliseconds `Sleep` slept by using `PaUtil_GetTime` like this:

```c
Sleep(next_sleep);
printf("[%d|%d],", next_sleep, (INT32)((PaUtil_GetTime() - start_time) * 1000));
```
It would show what is happening with timer in your application. In my experiment I got the following result:
```c
framesPerUser[ 180 ] framesPerHost[ 984 ] latency[ 22.31ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
[4|4],[0|0],[4|4],[0|0],[4|4],[0|0],[4|4],[0|0],[4|5],[0|0],[4|4],[0|0],[4|4],[0|0],[4|4],[0|0],[4|4],[0|0],[4|4],[0|0]
```

**Comment by kmatheussen** [comment]
I have a question about the WASAPI implementation. I haven't studied WASAPI very much, so please bear with me. In the WASAPI example from Microsoft, specifically in these two files:

https://github.com/microsoft/Windows-universal-samples/blob/master/Samples/WindowsAudioSession/cpp/WASAPIRenderer.h
https://github.com/microsoft/Windows-universal-samples/blob/master/Samples/WindowsAudioSession/cpp/WASAPIRenderer.cpp

They ask windows to call a method ("OnSampleReady") when it is possible to send data to WASAPI. Why can't PA do the same?

**Comment by dmitrykos** [comment]
@kmatheussen PA's implementation is following similar approach, please check `_PollGetOutputFramesAvailable`.

**Comment by kmatheussen** [comment]
@dmitrykos It doesn't seem like a similar approach to me. To me it seems like PA is checking now and then if it is possible to send data to WASAPI, while in the microsoft example, a callback is called by Windows when it's possible to send data to WASAPI.

**Comment by dmitrykos** [comment]
@kmatheussen we do not wait for notifications from WASAPI regarding ready samples in Poll mode (`paWinWasapiPolling`), instead Event mode is doing that (when `paWinWasapiPolling` is not set).

**Comment by kmatheussen** [comment]
Thank you. So what's the reason you might want to use using polling mode instead of event mode?

**Comment by kmatheussen** [comment]
Event mode only works in exclusive mode? (or is the same thing?)

**Comment by kmatheussen** [comment]
Another problem with the dynamic timing code above is that you don't measure the actual time WaitForSingleObject was using. Something like this should would fix it:

```
double last_time = 0;
double now = PaUtil_GetTime();

next_sleep -= (INT32)((now - last_time) * 1000);

INT32 half = next_sleep / 2;
if (half <= 2)
   next_sleep = 0;
else
   next_sleep = half;
last_time = now;
```

But the easiest and safest solution is probably just to always sleep 2ms, and not try to be smart about it. Even the above approach feels shaky.

**Comment by dechamps** [comment]
Just my two cents as a random bystander following this discussion: I'm really not sure using a fixed hardcoded 2ms wait time is a good idea. Keep in mind PortAudio is a general purpose library with many users, many applications, and many use cases. One of these use cases might be a music/video player that doesn't care about latency at all and uses large buffer sizes. Waking up every 2ms seems excessive in this case. It's not just a waste of CPU resources (although that makes me cringe in and of itself) - it can also be problematic for battery-powered devices, where battery efficiency tends to be significantly influenced by frequent wakeups. If we're going to wake up often, then at least make the polling frequency proportional to the buffer period.

**Comment by kmatheussen** [comment]
Again, that needs to be measured, but there's a good chance 2ms polling
won't have any practical affect on CPU usage or battery usage (no one cares
if your battery is drained 2 microseconds later than it otherwise would
have), and 2ms polling is already used for duplex and recording. Trying to
be smart about how much to wait is hard.

**Comment by dmitrykos** [comment]
@dechamps, there is no hardcoding to 2 ms in the current implementation for playback, for capturing there is a comment describing why it is needed:

```c
// WASAPI Input packets tend to expire very easily, let's limit sleep time to 2 milliseconds
// for all cases. Please propose better solution if any.
if (sleep_ms_in > 2)
    sleep_ms_in = 2;
```

@kmatheussen, Event mode works for Exclusive and Shared modes. I do not see any reason to measure overshoot of the sleep by `WaitForSingleObject` as my tests confirmed that it does not normally happen unless @sveinse finds out that it happens, then yes dynamic `next_sleep` calculation could take into account difference occurred due to imprecise sleep by `WaitForSingleObject` .

**Comment by sveinse** [comment]
@dmitrykos I'm going to test a little bit later, but a quick question: The only case I observe Event mode is if Exclusive mode is selected, otherwise Poll mode is used. There is a user flag to force it to Poll mode, but no way of specifying Event mode. How is the selection between Poll and Event mode made?

**Comment by kmatheussen** [comment]
Regarding trying to guess how much to sleep, there are uncertainties about if you have even hit the right point in time when it's best to send data to WASAPI (i.e. right after the buffer was sent to the sound card). If you don't hit this point exactly, then there will be a glitch some time into the future if the DSP code uses a little more than than usual if you try to guess the amount of time to sleep, and furthermore we can't really know for sure how WASAPI behaves when there are glitches, so this type of glitch might happen again and again.

In addition, I would be quite surprised if polling at 2ms, or even 1ms, would take up any significant amount of CPU, so that seems like the simplest and safest fix for this problem. Especially when portaudio doesn't even report glitches, although I hope that should be possible to fix.

**Comment by kmatheussen** [comment]
@sveinse https://www.cakewalk.com/Documentation?product=SONAR&language=3&help=AudioPerformance.29.html says that sonar only supports polling in shared mode in windows 7. But in windows 10 , sonar supports both polling and sharing at the same time.

**Comment by kmatheussen** [comment]
(The last part of the sentence should be: "supports event and sharing at the same time", not "supports polling and sharing at the same time".)

**Comment by sveinse** [comment]
@kmatheussen Interesting! A side note to this discussion, but the cakewalk document apparently hints at that it is possible to turn off Windows-side DSP processing from the WASAPI client. That I definitely want implemented in my own app, seeking as pristine output signal as possible.

**Comment by dmitrykos** [comment]
> The only case I observe Event mode is if Exclusive mode is selected, otherwise Poll mode is used

There was a problem with Event not working in Shared mode on Vista.

**Comment by sveinse** [comment]
@dmitrykos some test results:

The first test with separate `Sleep` fails with significant glitching.

```
WASAPI::OpenStream(output): framesPerUser[ 180 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
WASAPI: thread[ priority-0xD class-0x20 ]
[1|13],[1|15],[1|13],[1|14],[1|15],[1|15],[1|11],[1|14],[1|13],[1|13],[1|14],[1|14],[1|14],[1|14],[1|15],
```

```
WASAPI::OpenStream(output): framesPerUser[ 264 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
WASAPI: thread[ priority-0xD class-0x20 ]
[2|2],[2|11],[2|14],[2|15],[2|15],[2|15],[2|14],[2|14],[2|14],[2|14],[2|14],[2|14],[2|14],
```

```
WASAPI::OpenStream(output): framesPerUser[ 528 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
WASAPI: thread[ priority-0xD class-0x20 ]
[5|17],[5|14],[5|15],[5|14],[5|15],[5|15],[5|15],[5|15],[5|14],[5|14],[5|15],
```

In the second test, using `next_sleep` in `WaitForSingleObject()` also results in significant glitching.

```
WASAPI::OpenStream(output): framesPerUser[ 180 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
WASAPI: thread[ priority-0xD class-0x20 ]
[1|6],[1|14],[1|13],[1|15],[1|14],[1|14],[1|15],[1|15],[1|14],[1|14],
```

```
WASAPI::OpenStream(output): framesPerUser[ 264 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
WASAPI: thread[ priority-0xD class-0x20 ]
[2|6],[2|14],[2|12],[2|14],[2|15],[2|15],[2|15],[2|14],[2|15],[2|14],[2|15],[2|14],
```

```
WASAPI::OpenStream(output): framesPerUser[ 528 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
WASAPI: thread[ priority-0xD class-0x20 ]
[5|15],[5|15],[5|15],[5|14],[5|15],[5|15],[5|15],[5|15],[5|8],[5|13],[5|15],
```

Common to all of them is that the `Sleep` operation is orders of magnitude slower and late to return, and I think maybe this might be the root cause for the glitching. I've been working with other OSes that have somewhat the same property on sleep: Its always guaranteed that the process has slept at least `x` time, but with no guarantee for the upper bound of sleep. It could certainly look like `Sleep`  behaves like this.

I tried setting up the thread priority to `eThreadPriorityProAudio`: 

```
WASAPI::OpenStream(output): framesPerUser[ 528 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
WASAPI: thread[ priority-0xF class-0x20 ]
[5|17],[5|14],[5|15],[5|14],[5|14],[5|15],[5|15],[5|15],[5|15],[5|15],[5|14],
```

But that doesn't resolve anything in term of the sleep lengths.

Poll mode is currently the only method available for WASAPI in shared mode. So to test, I set `stream->out.streamFlags = AUDCLNT_STREAMFLAGS_EVENTCALLBACK;` and correspondingly for input to force Event mode. It does work on frame buffer sizes of >=528, but breaks down on lower values. And it segfaults if attempted on full duplex streams. -- Anyways, to keep focus on this topic, I'll stop and I will humbly suggest that we all defer this Poll/Event discussion in another new issue. Lets fix poll first.

**Comment by dmitrykos** [comment]
@sveinse your tests show that timer granularity of the process of your application is set to 15 ms and thus everything stopped working as it should. Most probably somewhere the `timeBeginPeriod(15)` was called?

Would you please put `timeBeginPeriod(1);` call into `ProcThreadPoll` somewhere in the beginning and try checking sleep times again. It looks that problem is not in PA but in granularity of Windows timer functions set to 15 ms somewhere inside your applicaton.

**Comment by sveinse** [comment]
Googling around and I find lots of references that Windows uses 15.6ms timer by default, which seems to match my observations. And I wouldn't be surprised that this were the case for the other users that are experiencing glitching in WASAPI. Using PA is realtime-y, and I think this warrants setting the system timer down to 1ms while PA is in active use in the poll thread. And to play nice with battery powered devices, set it back immediately as a stream is stopped. My 2 cents.

**Comment by dmitrykos** [comment]
@sveinse I think PA WASAPI also need to call `timeBeginPeriod()` to be sure that timer granularity is sufficient for timing operation. As per [this](http://download.microsoft.com/download/3/0/2/3027D574-C433-412A-A8B6-5E0A75D5B237/Timer-Resolution.docx) document 15 ms is standard Windows's granularity (!). 

So, please try adding `timeBeginPeriod(1);` call into `ProcThreadPoll` and if timing becomes as expected I will come up with the workaround using `timeBeginPeriod()` similar to DS and WDMKS implementations.

**Comment by sveinse** [comment]
@dmitrykos Yes, I'm sorry I wasn't clear: I have tested and verified that using `timeBeginPeriod(1)` in `ProcThreadPoll`

**Comment by dmitrykos** [comment]
@sveinse that's great, I will come with an implementation in form of merge request, so we could discuss it later on.

**Comment by kmatheussen** [comment]
Great work, but I'm pretty sure the code is broken in other ways too. If you're trying to save CPU cycles and battery power by not polling very frequently (for instance 2ms), at least you have to synchronize the processing thread to WASAPI. I think something like this is required for correct operation:
```
for(;;){
  while(wasapi_not_ready()){
     Sleep(1); // should normally be called 0, 1, or 2 times.
  }
  time = get_current_time();
  data = process_audio();
  wasapi_send_data(data);
  duration = get_current_time() - time;
  Sleep(duration_in_ms_of_one_audio_period - 2);
}
```

In addition, are you sure you shouldn't process audio before waiting for WASAPI to be ready, and not the other way around which is the case now? It may require allocating another buffer, but it sounds safer.

**Comment by kmatheussen** [comment]
The second laste line in the code should look like this:
```
Sleep(duration_in_ms_of_one_audio_period - duration - 2);
```
Or something like this at least.

**Comment by sveinse** [comment]
@kmatheussen if I understand you last comment correctly: WASAPI is actually preloading data. The first processing iteration of the audio data is done here:

https://github.com/PortAudio/portaudio/blob/master/src/hostapi/wasapi/pa_win_wasapi.c#L5829-L5877

On low buffer sizes/latencies, the timer granularity of 1 ms becomes very coarse. Given the inherent jitter in OS Sleep I agree that is its far better to sleep short than to have an exact sleep that jitters around too early and too late. Sleeping too long eats at the time available for the user callback.

You probably already know this, but I think this Wiki entry is a good explanation:
https://github.com/PortAudio/portaudio/wiki/BufferingLatencyAndTimingImplementationGuidelines#basic-bufferingtiming-model

Polling as a methodology for asynch communication always have these types drawbacks in terms of how to deal with sleep and how often to poll for data. In my opinion the event based method would be better because the thread is only waken up when data is available and/or requested. However, the current event implementation of wasapi is not available for streams except for exclusive mode and it fails to work with lower buffer sizes. I believe these things should be looked at too, but as I said, let's take one item at a time.

**Comment by dmitrykos** [comment]
@sveinse I created a pool request #307 with discussed changes. Probably due to transition to GitHub pull request gathered some earlier commits which were merged earlier, nevertheless the last one is the one we are discussing. Please check if it works fine for your application.

It will set system timer granularity to 1 ms and then revert it to default when Poll thread function exits. Also the implementation is using the proposed mechanism to subtract the processing time from the next sleep time as it allows to improve the periodicity of the polling time slots.

**Comment by dmitrykos** [comment]
@sveinse I created a pool request #307 with discussed changes. Probably due to transition to GitHub pull request gathered some earlier commits which were merged earlier, nevertheless the last one is the one we are discussing. Please check if it works fine for your application.

It will set system timer granularity to 1 ms and then revert it to default when Poll thread function exits. Also the implementation is using the proposed mechanism to subtract the processing time from the next sleep time as it allows to improve the periodicity of the polling time slots.

**Comment by sveinse** [comment]
@dmitrykos Tested the PR and it glitches unfortunately. It is particularly evident on frame sizes such as 528 and 1056, and even higher values glitches occasionally. The lower values (264, 132, 66) are stable, but there is a train of glitching in the very beginning, so I suspect there are glitching issues in the first iteration / warm-up.

On the positive note, I can confirm that the `timeBeginPeriod()` logic works. I confirm that it starts with a high value/default, sets it to 1, runs the loop and restores the previous value when done.

**Comment by dmitrykos** [comment]
@sveinse thank you very much, I am surprised but ok, let's find out what is the problem now. In the new implementation I removed sleep time division by 2, that is the only change, the rest of the logic remains the same.

Would you please replace `GetNextSleepTime` to such:

```c
static inline INT32 GetNextSleepTime(SystemTimer *timer, ThreadIdleScheduler *scheduler, LONGLONG startTimeMs, 
	UINT32 sleepMs)
{
	INT32 nextSleepMs;

	// Get next sleep time
	if (sleepMs == 0)
		nextSleepMs = ThreadIdleScheduler_NextSleep(scheduler);
	else
		nextSleepMs = sleepMs;

	// Adjust next sleep time dynamically depending on how much time was spent in ProcessOutputBuffer/ProcessInputBuffer
	// therefore periodicity will not jitter or be increased for the amount of time spent in processing;
	// [ssss|p],[ssss|pp],[sss|p],[ssss|ppp],[s|p],[ssss|pppp],[p],[ssss|p]
	INT32 procTime = (INT32)(SystemTimer_GetTimeMsec(timer) - startTimeMs);
	printf("{%d},", procTime);
	nextSleepMs -= procTime;
	if (nextSleepMs < 0)
		nextSleepMs = 0;

	return nextSleepMs;
}
````

and inside the processing loop to such code:

```c
if (!PA_WASAPI__IS_FULLDUPLEX(stream))
{
    // Processing Loop
    while (1)
    {
        LONGLONG startBeforeWait = SystemTimer_GetTimeMsec(&timer);

        if (WaitForSingleObject(stream->hCloseRequest, nextSleepMs) != WAIT_TIMEOUT)
            break;

        startTimeMs = SystemTimer_GetTimeMsec(&timer);

        printf("[%d|%d],", nextSleepMs, (INT32)(startTimeMs - startBeforeWait));

        for (i = 0; i < S_COUNT; ++i)
```

`printf` will help to see the processing and polling time slots. In my test I got the following result:

```c
framesPerUser[ 480 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]

[9|9],{2},[8|8],{1},[9|9],{1},[9|9],{3},[7|7],{2},[8|8],{3},[7|7],{2},[8|8],{2},[8|8],{3},[7|7],{2},[8|8]
```

Inside [] we can see the polling time, inside {} the processing time, so I hope we can see what is happening in your case.

**Comment by sveinse** [comment]
@dmitrykos 

Glitching during start, some intermittent glitching during this 1 second playback:

```
WASAPI::OpenStream(output): framesPerUser[ 1056 ] framesPerHost[ 2112 ] latency[ 44.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
[22|22],{0},[22|22],{0},[22|23],{0},[22|23],{0},[22|23],{0},[22|22],{0},[22|22],{0},[22|23],{1},[21|21],{0},[22|21],{1},[21|20],{1},[21|21],{1},[21|20],{1},[21|20],{1},[21|21],{1},[21|21],{1},[21|21],{1},[21|21],{0},
[22|22],{1},[21|20],{1},[21|21],{0},[22|22],{0},[22|23],{1},[21|21],{0},[22|23],{0},[22|21],{0},[22|22],{0},[22|23],{1},[21|21],{0},[22|23],{0},[22|22],{1},[21|21],{1},[21|21],{0},[22|22],{0},[22|22],{1},[21|21],{1},
[21|22],{1},[21|20],{0},[22|23],{0},[22|23],{0},[22|21],{0},[22|22],{0},[22|23],{0},[22|22],{0},[22|21],{0},
````

More glitching:

```
WASAPI::OpenStream(output): framesPerUser[ 528 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
WASAPI: thread[ priority-0xD class-0x20 ]
[11|12],{0},[11|11],{0},[11|12],{28},[0|0],{1},[10|10],{0},[11|11],{0},[11|12],{0},[11|11],{0},[11|10],{0},[11|11],{0},[11|11],{0},[11|12],{0},[11|11],{0},[11|10],{0},[11|11],{1},[10|10],{1},[10|9],{1},[10|10],{0},[11|12],{1},[10|10],{0},[11|11],{1},[10|10],{0},[11|11],{0},[11|11],{0},[11|11],{0},[11|12],{0},[11|12],{0},[11|11],{0},[11|11],{0},[11|11],{0},[11|11],{1},[10|10],{0},[11|11],{1},[10|10],{0},[11|12],{0},[11|11],{0},[11|11],{1},[10|10],{0},[11|12],{0},[11|11],{0},[11|11],{0},[11|10],{1},[10|10],{0},[11|11],{0},[11|11],{1},[10|10],{0},[11|11],{0},[11|11],{0},[11|10],{1},[10|10],{0},[11|12],{1},[10|10],{0},[11|10],{0},[11|12],{0},[11|11],{1},[10|10],{0},[11|11],{0},[11|11],{0},[11|10],{0},[11|12],{0},[11|11],{0},[11|11],{0},[11|12],{0},[11|12],{0},[11|11],{0},[11|11],{0},[11|11],{1},[10|10],{0},[11|11],{0},[11|11],{1},[10|10],{0},[11|12],{0},[11|12],{0},[11|11],{1},[10|10],{0},[11|12],{0},[11|10],{0},[11|11],{0},[11|11],{0},[11|11],{0},[11|10],{0},[11|12],{0},[11|11],{1},[10|10],{0},[11|11],{1},[10|10],{0},[11|10],{0},[11|11],{1},[10|9],{1},[10|10],{0},[11|11],{0},
```

**Comment by dmitrykos** [comment]
@sveinse i looks that processing is not the problem (1-2 ms), there was spike though `{28}` but still it is within the time slot of the host buffer.

Could you apply the change which you proposed inside the `ConfigureLoopSleepTimeAndScheduler`, e.g. `sleepMsecOut /= 2` that you proposed earlier, just to see if it affects glitching. If it is then most probably when we are in situation when 1/2 of the host buffer equals the user buffer we should wait less, e.g. 1/4 of the host buffer. It is interesting to check this assumption.

**Comment by dmitrykos** [comment]
By the way I just got glitching for the first time when set latency to 11 ms:

```c
[11|11],{1},[11|11],{1},[11|11],{1},[11|12],{1},[11|12],{1},[11|11],{2},[10|10],{2},[10|11],{1},[11|10],{1},[11|11]
```

**Comment by sveinse** [comment]
@dmitrykos I tested with /2 at the glitching goes away, except a slight glitch at 264 (host is 1056) at startup.

It struck me that the audio device is being spun up from no audio when I tested this, so I ran another series of tests with audio playing in the audio device prior and during the test. This was to ensure that the Windows WSAPI layer and WDM driver is running and active. This improved startup stability/glitching a lot. Lower frame buffer sizes, 264 and less can glitch occasionally, but significantly improved if and when the audio device is already running.

**Comment by kmatheussen** [comment]
As long as you don't synchronize your thread to WASAPI, this will never
work. You have to do it the way I roughly described it in the pseudocode,
but even then you should probably monitor the duration since last time you
sent data to wasapi in case sleeping takes longer time than usual, or
anything else unusual happens.

As a first step though, I think it would be best to set timing to 2ms for
everyone, and push it, so that pulseaudio starts working again. Step two
would be to let users of windows 10 (and perhaps windows 8) use event-based
instead. Users of windows 7 and earlier are probably used to bad
performance in general anyway, so they might not notice a possible battery
drain or high CPU usage of polling constantly at 2ms.

**Comment by sveinse** [comment]
@kmatheussen 
> As long as you don't synchronize your thread to WASAPI, this will never work. 

Out of curiosity, "your thread" then you mean then PA poll thread, not any user application thread, right?

On the user side, I run all processing in the context of the PA callback, which the thread being created specific to the selected host API. My app is a python app (!). When I started this project I expected I had to do the audio processing in a separate user thread, but for most host apis (all except wasapi), it actually works great as is.

**Comment by kmatheussen** [comment]
On Thu, Oct 1, 2020 at 10:40 AM Svein Seldal <notifications@github.com>
wrote:

> @kmatheussen <https://github.com/kmatheussen>
>
> As long as you don't synchronize your thread to WASAPI, this will never
> work.
>
> Out of curiosity, "your thread" then you mean then PA poll thread, not any
> user application thread, right?
>

Right, the while loop in the "ProcThreadPoll" thread in pa_win_wasapi.c
must be synchronized to wasapi. If you don't synchronize it will be
coincidental if you hit the right moment to send data to wasapi. Even if
processing audio doesn't take any time, the while loop timer is likely to
drift away from the soundcard timer so that you will get glitches now and
then.

**Comment by dmitrykos** [comment]
@sveinse I committed an update which will use your proposed workaround (it is in `GetSleepTime` function), I still decided to be conservative and apply this workaround if user buffer is 1/2 of the host buffer. 

I did number of listening tests using 440 Hz sine wave at 48 kHz:

```c
framesPerUser[ 96 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
[2|2],{2},[0|0],{0},[2|2],{0},[2|2],{0},[2|2],{0},[2|3],{1},[1|1],{0},[2|3],{0},[2|2],{0},[2|2],{1},[1|2],{0},[2|2]

framesPerUser[ 152 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
[3|4],{0},[3|4],{2},[1|1],{0},[3|4],{0},[3|2],{1},[2|2],{1},[2|1],{1},[2|2],{0},[3|4],{2},[1|0],{0},[3|4],{1},[2|1]

framesPerUser[ 192 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
[4|5],{0},[4|5],{0},[4|4],{1},[3|2],{0},[4|5],{0},[4|5],{1},[3|2],{0},[4|4],{1},[3|3],{0},[4|4],{0},[4|4],{1},[3|2]

framesPerUser[ 240 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
[5|5],{1},[4|4],{0},[5|5],{1},[4|4],{0},[5|5],{1},[4|3],{0},[5|6],{1},[4|4],{0},[5|4],{1},[4|4],{0},[5|6],{1},[4|3]

framesPerUser[ 300 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
[6|6],{1},[5|4],{0},[6|6],{0},[6|7],{1},[5|5],{0},[6|7],{0},[6|7],{1},[5|4],{0},[6|6],{1},[5|4],{0},[6|7],{0},[6|7]

framesPerUser[ 376 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
[3|4],{0},[3|3],{0},[3|4],{0},[3|4],{0},[3|4],{0},[3|3],{0},[3|3],{0},[3|4],{0},[3|3],{0},[3|4],{0},[3|4],{0},[3|4]

framesPerUser[ 384 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
[4|4],{1},[3|2],{0},[4|5],{0},[4|5],{1},[3|2],{0},[4|3],{1},[3|2],{0},[4|5],{0},[4|5],{1},[3|3],{0},[4|4],{1},[3|3]

framesPerUser[ 480 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
[5|5],{0},[5|5],{1},[4|4],{0},[5|5],{1},[4|4],{0},[5|6],{1},[4|3],{0},[5|5],{1},[4|3],{0},[5|6],{0},[5|6],{0},[5|6]
```

Logic is performing well, time slots are stable and dynamically adjusted depending on processing. Would you please check on your application how it works and if stable. If glitching persist then we could make `sleepTimeOut /= 2` unconditional.

**Comment by dmitrykos** [comment]
> it will be coincidental if you hit the right moment to send data to wasapi

Actually in our application we really hit the right moment, in the worst case scenario we are filling 50% of the host buffer while OS is sending another 50% to the hardware. With the last update to the pull request we will be filling 25% of the host buffer while WASAPI is sending 75% of the host buffer to hardware. In case of @sveinse it helps to remove glitching/underruns which happen not on PA's level (it is confirmed by time slots logging) but somewhere in WASAPI/driver level.

**Comment by kmatheussen** [comment]
On Thu, Oct 1, 2020 at 11:03 AM Dmitry Kostjuchenko
<notifications@github.com> wrote:
>
> it will be coincidental if you hit the right moment to send data to wasapi
>
> Actually in our application we really hit the right moment, in the worst case scenario we are filling 50% of the host buffer while OS is sending another 50% to the hardware. With the last update to the pull request we will be filling 25% of the host buffer while WASAPI is sending 75% of the host buffer to hardware. In case of @sveinse it helps to remove glitching/underruns which happen not on PA's level (it is confirmed by time slots logging) but somewhere in WASAPI/driver level.

Oh, okay. I see you have changed your code. I guess this code:

if (framesAvail < framesProc)
{
  nextSleepMs = 0;
  continue;
}

at https://github.com/PortAudio/portaudio/blob/1e4ee2003d20e41792a4cd25cb88320da0c85105/src/hostapi/wasapi/pa_win_wasapi.c#L6059
in practice works as a synchronization mechanism now?

**Comment by dmitrykos** [comment]
@kmatheussen this piece of code is workaround for another case, I provided explanation in the comment above it.

**Comment by kmatheussen** [comment]
On Thu, Oct 1, 2020 at 11:23 AM Kjetil Matheussen
<k.s.matheussen@gmail.com> wrote:
>
> On Thu, Oct 1, 2020 at 11:03 AM Dmitry Kostjuchenko
> <notifications@github.com> wrote:
> >
> > it will be coincidental if you hit the right moment to send data to wasapi
> >
> > Actually in our application we really hit the right moment, in the worst case scenario we are filling 50% of the host buffer while OS is sending another 50% to the hardware. With the last update to the pull request we will be filling 25% of the host buffer while WASAPI is sending 75% of the host buffer to hardware. In case of @sveinse it helps to remove glitching/underruns which happen not on PA's level (it is confirmed by time slots logging) but somewhere in WASAPI/driver level.
>
> Oh, okay. I see you have changed your code. I guess this code:
>
> if (framesAvail < framesProc)
> {
>   nextSleepMs = 0;
>   continue;
> }
>
> at https://github.com/PortAudio/portaudio/blob/1e4ee2003d20e41792a4cd25cb88320da0c85105/src/hostapi/wasapi/pa_win_wasapi.c#L6059
> in practice works as a synchronization mechanism now?

Yes, that's what happens. The code looks much better now, but as far
as I can see there's a big risk that you will always spend more time
inside one iteration of the while loop than the duration of one wasapi
buffer, slowly causing the point of time for providing samples to
wasapi to drift later and later. This:

- nextSleepMs -= (INT32)(SystemTimer_GetTimeMsec(timer) - startTimeMs);
+ nextSleepMs -= (INT32)(SystemTimer_GetTimeMsec(timer) - startTimeMs - 2);

might fix that for most situations, but I have a feeling that it might
not be good enough for all situations.

Or perhaps I'm not looking at the latest version of the source code?

**Comment by kmatheussen** [comment]
On Thu, Oct 1, 2020 at 11:49 AM Dmitry Kostjuchenko <
notifications@github.com> wrote:

> @kmatheussen <https://github.com/kmatheussen> this piece of code is
> workaround for another case, I provided explanation in the comment above it.
>

Okay, but it does provide some synchronization as well.

**Comment by dmitrykos** [comment]
@kmatheussen there is no drifting because we always query number of available frames and try to fill them with `_PollGetOutputFramesAvailable`. Host buffer may become smaller or a bit bigger but will not drift constantly to any direction.

**Comment by kmatheussen** [comment]
On Thu, Oct 1, 2020 at 11:54 AM Dmitry Kostjuchenko <
notifications@github.com> wrote:

> @kmatheussen <https://github.com/kmatheussen> there is no drifting
> because we always query number of available frames and try to fill them
> with _PollGetOutputFramesAvailable. Host buffer may become smaller or a
> bit bigger but will not drift constantly to any direction.
>

Sorry, bear with me. Hope you will explain it to me. If you're right,
great. If I'm right, that's great too.

It looks like you are sleeping the entire buffer. This is the line that
calculates how long time to sleep (the actual time to process audio is
subtracted from this number at each iteration):

sleepMsecOut =
GetFramesSleepTimeMicroseconds((stream->bufferProcessor.framesPerUserBuffer
? stream->bufferProcessor.framesPerUserBuffer :
stream->out.params.frames_per_buffer),
stream->out.wavex.Format.nSamplesPerSec);

How can you avoid risking drifting when you don't even keep account of how
long you have slept totally (every iteration of the while loop you sleep
the maximum amount of time). Is the WASAPI buffer bigger than the PA
buffer? Can _PollGetOutputFramesAvailable return a negative number?

**Comment by kmatheussen** [comment]
(Wrong line.
sleepMsecOut = GetFramesSleepTime(stream->out.framesPerBuffer,
stream->out.wavex.Format.nSamplesPerSec);
)

**Comment by sveinse** [comment]
@dmitrykos Sorry for the delay, I got put off by the rebase in winrt and had to remerge my changes (I have to patch msvc build files to be able to build it).

Running the PR #307 as is, with `PA_WASAPI_LOG_TIME_SLOTS` enabled, seem to create decent playback on my system in all buffer sizes. There is still some glitching in the very beginning of frame sizes around 264.

```
WASAPI::OpenStream(output): framesPerUser[ 264 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
[0|1],{0},[5|4],{1},[4|4],{1},[4|3],{1},[4|3],{1},[4|4],{1},[4|4],{1},[4|4],{0},[5|5],{0},[5|5],{0},[5|5],{0},[5|5],{0},[5|6],{0},
```

You mentoned you'd like to test my app on your HW. I created this special build of ELNS using the portaudio version from this PR. It can be downloaded from https://github.com/sveinse/elns-portaudio-test/releases/tag/1. It has console output enabled, so running it from a command-prompt will output the WASAPI logging. This test version loads the input file into memory prior to playback to rule out any timing issues internally.

I test by pressing play and stop rather quickly. What is interesting to note is that it never glitches on rapid pause play operations. Perhaps that suggests that this is related to the initial startup of a stream?

**Comment by sveinse** [comment]
@dmitrykos btw, to test your statement in PR #307 "...time hungry user-side processing", I added the following code:

```C
		LARGE_INTEGER a, b;
		QueryPerformanceCounter(&b);
		processor[S_OUTPUT].processor(NULL, 0, data, frames, processor[S_OUTPUT].userData);
		QueryPerformanceCounter(&a);
		printf("_%lld_,", a.QuadPart-b.QuadPart);
```

```
WASAPI::OpenStream(output): framesPerUser[ 264 ] framesPerHost[ 1056 ] latency[ 22.00ms ] exclusive[ NO ] wow64_fix[ NO ] mode[ POLL ]
_932_,[0|1],_2312_,_575_,_1512_,_490_,{1},[4|4],{0},[5|6],_979_,{0},[5|5],{0},[5|5],_1013_,_644_,{1},[4|3],{0},[5|5],_1004_,_646_,{1},[4|3],{0},[5|6],_1012_,_624_,{1},[4|4],{0},[5|6],_1204_,_696_,{0},[5|6],{0},[5|5],_1023_,{1},[4|4],{0},[5|5],_1113_,_732_,{1},[4|3],{0},[5|4],_1310_,_1326_,{1},[4|4],{0},[5|5],_1276_,_936_,{1},[4|3],{0},[5|6],_1402_,_884_,
```

This one is a glitch on startup. It can seem like the user callback spends much longer time in the first invocation, but the number are very variable. They are equally variable and high on playbacks that doesn't glitch and even with lower buffer sizes. I plan to collect a better histogram of these, but just wanted to mention that I'm looking into if the root cause to the startup glitching is excessive user callback times.

**Comment by kmatheussen** [comment]
On Thu, Oct 1, 2020 at 8:03 PM Dmitry Kostjuchenko <notifications@github.com>
wrote:

> @kmatheussen <https://github.com/kmatheussen>
>
> It looks like you are sleeping the entire buffer.
>
> We are sleeping for the duration of the user buffer which is always at
> least 2x times smaller than host (WASAPI) buffer, if user specifies the
> desired latency the host buffer will be of the size of the requested
> latency. There is no mistake in the implementation.
>

Thanks for clearing that up Dmitry. I suggest adding this line "// the user
buffer which is always at least 2x times smaller than host (WASAPI) buffer"
to the code.

**Comment by dmitrykos** [comment]
@sveinse thank you for a test and I am glad that we managed to get to a stable (almost) version. 

> with `PA_WASAPI_LOG_TIME_SLOTS` enabled

I probably misunderstood but `PA_WASAPI_LOG_TIME_SLOTS` needs to be defined only to enable logging (printfs). Without defining it I believe you have the same stable playback?

I could not test your application unfortunately as it throws `Invalid device` error, tried on 2 machines but with the same result. Audio device is set to 24/48000 format. May be I am doing something wrong. Input is set to WAV/FLAC file, Output to WASAPI output device.

```c
Starting
Using input device index number: -1
Using output device index number: 4
WASAPI: IAudioClient2 set properties: IsOffload = 0, Category = 0, Options = 0
WASAPI ERROR HRESULT: 0x88890008 : AUDCLNT_E_UNSUPPORTED_FORMAT
 [FUNCTION: CreateAudioClient FILE: \elns-wasapi\portaudio-pr307\src\hostapi\wasapi\pa_win_wasapi.c {LINE: 3132}]
WASAPI ERROR PAERROR: -9997 : Invalid sample rate
 [FUNCTION: ActivateAudioClientOutput FILE: \elns-wasapi\portaudio-pr307\src\hostapi\wasapi\pa_win_wasapi.c {LINE: 3449}]
WASAPI ERROR PAERROR: -9996 : Invalid device
 [FUNCTION: OpenStream FILE: \elns-wasapi\portaudio-pr307\src\hostapi\wasapi\pa_win_wasapi.c {LINE: 2143}]
An error occured while using the portaudio stream
Error number: -9996
Error message: Invalid device
Traceback:
  File "elns\core.py", line 797, in tick_step
  File "elns\audio.py", line 556, in open
  File "pyaudio.py", line 754, in open
  File "pyaudio.py", line 445, in __init__
OSError: [Errno -9996] Invalid device
ERROR: Opening audio device failed. OSError: [Errno -9996] Invalid device
```

**Comment by sveinse** [comment]
@dmitrykos This is actually another bug (or feature) in WASAPI. The root cause is that you're most probably trying to play an audio file with a different sample rate than the Windows audio device is configured for (48k). In this version I do not use the `paWinWasapiAutoConvert` flag, so the sample rate must match. WASAPI first returns error -9997 Invalid sample rate which is an useful error. But it later mutates into a -9996 Invalid device which is the message that is reported to the user application. Unfortunately this second message is more cryptic that the first and is confusing at best.

**Comment by dmitrykos** [comment]
@sveinse yes, you were right, matching sample rate made it work (I tried to match them yesterday but may be something went wrong) and I went through all the latencies. I could hear some very occasional glitching when using 128 ms but you enabled logging with `PA_WASAPI_LOG_TIME_SLOTS` and then may be console I/O due to a lot of `printf` could cause it. All other latencies were stable when listening to 440 Hz sine tone.

Can we conclude that we managed to reach stability and current implementation shall be merged, or you wish to make some other experiments first?

**Comment by sveinse** [comment]
@dmitrykos Please stand by, I'm going to run some test now without printf.

> > This is actually another bug (or feature) in WASAPI
> 
> It is not a bug but how WASAPI works in Shared mode, Microsoft decided to pass the resampling task to the application level.

The sample rate matching error is correct. That's not what i meant. I meant that WASAPI does identify this being a sample rate error (the -9997 error), but it later becomes a -9996 invalid device error, where the latter is the error reported back to the user layer. It would be better if the error was -9997 Invalid sample rate.

**Comment by dmitrykos** [comment]
@sveinse I see, sorry for confusion. I check the related code and it is actually a bug in the implementation at line 3842 `if (FAILED(hr = ActivateAudioClientOutput(stream)))`. The same for input. I propose to fix it with another pull request once we are done with current one.

**Comment by philburk** [comment]
This is a great technical discussion. But the thread is very long. So when this issue is closed, could someone please add a very short summary paragraph describing the issue and the final solution. That will help others in the future. Thanks.

**Comment by sveinse** [comment]
@dmitrykos I rebuild portaudio to release with no printfs. Unfortunately the glitching does not go away

The test results are as follows:

Devices:
```
   1: Windows WASAPI  [out 2]  LG HDR 4K (NVIDIA High Definition Audio)
   2: Windows WASAPI  [out 2]  Speaker/HP (Realtek High Definition Audio)
   3: Windows WASAPI  [out 2]  Speakers (2- Logitech USB Headset)
```

Solo = started without any audio running on that device. shared = another audio playback running on the same device

| buffer size | Device 1 solo | Device 1 shared | Device 2 solo | Device 2 shared | Device 3 solo |
| --- | --- | --- | --- | --- | --- |
| 2112 | 3 OK | 3 OK | 3 OK | 5 OK | 3 OK |
| 1056 | 3 OK | 2 OK, 1 glitch | 3 OK | 5 OK | 3 OK, 2 glitch |
| 528 | 1 OK, 5 glitch | 5 glitch | 3 OK , 3 glitch | 5 glitch | 4 glitch |
| 264 | 4 glitch | 3 OK, 2 glitch | 1 OK, 4 glitch | 2 OK, 3 glitch | 3 glitch |
| 132 | 5 glitch | 3 OK, 2 glitch | 2 OK, 3 glitch | 5 OK | 4 OK |
| 66 | 5 glitch | 4 OK | 1 OK, 4 glitch | 5 OK | 3 OK, 1 glitch |

**Comment by dmitrykos** [comment]
@sveinse thank you for a very detailed test! Did it become worse when you do not define `PA_WASAPI_LOG_TIME_SLOTS`?

**Comment by sveinse** [comment]
@dmitrykos 

> @sveinse thank you for a very detailed test! Did it become worse when you do not define `PA_WASAPI_LOG_TIME_SLOTS`?

I changes multiple variables in the test above (no printf and Release build), so I had to rerun the tests. Setting `PA_WASAPI_LOG_TIME_SLOTS` doesn't improve glitching. What is worth noting is that the glitch sound coming at a specific buffer value are more or less deterministic. E.g. at 528 there is a glitch at the startup when it glitches, nothing more. At 264 when it isn't perfect playback, I get a double glitch at the beginning with subjectively equal distance. So I would be tempted to think that the system is behaving close to deterministic when it glitches.

> According to the time slots logging results, on application/PA side all seems to be fine and we never exceed 1/2 of the host buffer latency, so I am puzzled.

Is it Windows that require a larger buffer fill in the first invocations to be able to handle internal warm-up?

I am considering logging/tracing the absolute time when e.g. `WasapiHostProcessingLoop()` is called or somesuch to see if there is correlation in the timing that might explain the glitching.

The other worry I have is that these glitches are unobserved. There is no flag being raised and sent back to the user indicating the glitch/underrun. That bothers me.

> edit:
> Comparing other APIs does not makes sense as in one case core (DS, MME) may be doing some internal buffering which works fine, WDMKS instead is a direct I/O without any hidden wrapper. We may be facing some corrupted logic of WASAPI's shared mode. For experiment, to check if PA WASAPI logic is correct you could try forcing Exclusive mode which bypasses WASAPI's mixer/wrapper, if playback is stable then most probably there is something wrong with shared mode of WASAPI then.

I understand your point, but at the same time I think its not good if the WASAPI host API underperforms compared to the other host APIs. So my opinion and objective is that we need to find a config and method that works reliably for the users. If that mode is Event mode, that's fine, except the current Event mode is unavailable except for Exclusive mode.

Exclusive mode, which use Event, is rock solid on all frame size I can throw at it, even with very low values such as 12 samples. One question: In shared mode I see that the host buffer sizes orient around 1056 samples in my system. I cannot find any such "converging" values in exclusive mode. It seems to want to use a host buffer size a couple of hundred samples bigger than the user buffer size. Why is that?

**Comment by dmitrykos** [comment]
@sveinse 
> Is it Windows that require a larger buffer fill in the first invocations to be able to handle internal warm-up?

There are no docs related to how many frames should be preloaded to guarantee stable startup but there is headroom for improvement by filling whole host buffer with user data before stream is started:

you could modify block starting at line line 5992 to such:
```c
while (frames >= stream->out.framesPerBuffer)
{
    if ((hr = ProcessOutputBuffer(stream, processor, stream->out.framesPerBuffer)) != S_OK)
    {
        LogHostError(hr); // not fatal, just log
        break;
    }
}
```

> The other worry I have is that these glitches are unobserved. There is no flag being raised and sent back to the user indicating the glitch/underrun. That bothers me.

Yes, it may be the case that problem is somewhere inside the WASAPI's mixer/buffer handling because the time slots logging showed that no underruns happening and polling time is always smaller than 1/2 of the host buffer. This problem is happening when the lowest (22 ms) possible host buffer is used in Shared mode. In your tests you got stable playback when the host buffer has 44 ms latency (2112).

> my opinion and objective is that we need to find a config and method that works reliably for the users

I agree but here to my view we can not compare performance 1 to 1, for example using your application and selecting DS as host API I can see in logs that DS latency is much higher than what we are trying to achieve with WASAPI:

`DirectSound host buffer size frames: 5548, polling period seconds: 0.007483, @ sr: 44100.000000`

e.g. it is 125 ms when I selected latency 256. MME's latency will be high too and WDMKS is equivalent of WASAPI's Exclusive mode. On contrary in case of WASAPI host API we are trying to play with 22 ms latency in Shared mode. May be it is too low for Shared mode while 44 ms becomes stable which is almost 3 times lower than DS's latency anyway.

If you could experiment - force Exclusive mode for WASAPI and test the range of latencies then we could understand whether Shared mode itself is buggy/unstable when latency is 22 ms or the problem in PA WASAPI implementation (I can see no corrupted logic in our implementation). If playback is stable then we probably can not achieve an absolute success with WASAPI Shred mode with latency below 44 ms.

> One question: In shared mode I see that the host buffer sizes orient around 1056 samples in my system.

It is the value provided from WASAPI's API by `IAudioClient_GetBufferSize`, so rationale behind it is in WASAPI's implementation code.

**Comment by sveinse** [comment]
@dmitrykos I did not do a very thorough test, but I get no glitching with any of the buffer size values I've tested before. Looking good. This test was in debug and with printf, so let me do a more detailed test.

**Comment by sveinse** [comment]
@dmitrykos I have now tested the latest proposal extensively:

**Shared mode:** Works perfectly without any glitching. Tested with 3 different audio devices (built-in, HDMI and USB), in solo and mixed with other playback devices. And tested every permutations with buffer sizes of 2112, 1056, 528, 264, 132 and 66. All OK.

**Exclusive mode:** Works great, but I get some glitching on the lower buffer sizes (132 and below). This is probably due to excessive user processing time and will at some point be expected.

**Forced event mode:** In this mode I force the playback to use the Event mode. It is not stable at all except for 528. In all the other values it creates a repetitive, monotonic stream of glitching audio (as opposed to random, intermittent glitching).

**Auto convert mode:** Didn't do as detailed test as above, but using this option doesn't seem to affect stability. If I revert the fix above, I get the same type of glitching regardless if the auto convert feature is used or not.

> I agree but here to my view we can not compare performance 1 to 1, for example using your application and selecting DS as host API I can see in logs that DS latency is much higher than what we are trying to achieve with WASAPI:
> 
> `DirectSound host buffer size frames: 5548, polling period seconds: 0.007483, @ sr: 44100.000000`
> 
> e.g. it is 125 ms when I selected latency 256. MME's latency will be high too and WDMKS is equivalent of WASAPI's Exclusive mode. On contrary in case of WASAPI host API we are trying to play with 22 ms latency in Shared mode. May be it is too low for Shared mode while 44 ms becomes stable which is almost 3 times lower than DS's latency anyway.

I agree, DS and MME is not a fair comparison due to their very long buffering schemes and that they can't be compared 1 to 1. However my overall point is seen from the user's perspective: The user most often don't care which host API is used, just something that works. DS and MME are old legacy APIs (which is using WASAPI behind the scene). I would like to be able to recommend my users to use WASAPI as this is preferred, modern, Windows audio system. That's all.

**Comment by sveinse** [comment]
PS! I hit the bug/feature in https://github.com/PortAudio/portaudio/blob/winrt/src/hostapi/wasapi/pa_win_wasapi.c#L2079-L2085 . Long story short: I have an Focusrite Scarlett 6i6 USB Pro sound card, and it wanted a firmware update. For some reason that firmware update completely broke audio on my computer. If I selected that audio card as output, other audio apps, such as Spotify crashed. I had plans to use this sound card for the tests above, but had to give up. When I ran my application, PortAudio crashed due to this:

```
Cannot create DirectSound for Speakers (Focusrite Usb Audio). Result = 0x88890008
Speakers (Focusrite Usb Audio)'s GUID: {0xd802d747,0x1f0a,0x47f8,0x98,0xf9,0x47,0xfe,0x1b,0xa3,0x2f, 0x35}
....
WASAPI: device idx: 03
WASAPI: ---------------
WASAPI:3| name[Speakers (Focusrite Usb Audio)]
WASAPI:3| form-factor[1]
WASAPI ERROR HRESULT: 0x88890008 : AUDCLNT_E_UNSUPPORTED_FORMAT
 [FUNCTION: CreateDeviceList FILE: C:\Svein\Prosjekt\elns-wasapi\portaudio-pr307\src\hostapi\wasapi\pa_win_wasapi.c {LINE: 2084}]
WASAPI: failed CreateDeviceList error[-9986|Internal PortAudio error]
WASAPI: failed PaWasapi_Initialize error[-9986|Internal PortAudio error]
```

This is OT for this issue, but I wanted to mention it. If you like I could create a new issue for it.

**Comment by dmitrykos** [comment]
@sveinse I am glad to hear we got some light in the end of tunnel. Do I understand it correctly that proposed modification starting from line 5992 improved situation with glitches?

> This is OT for this issue, but I wanted to mention it. If you like I could create a new issue for it.

Yes, if you do not mind to create 2 issues, one for this issue and another for `The sample rate matching error is correct. That's not what i meant. I meant that WASAPI does identify this being a sample rate error (the -9997 error), but it later becomes a -9996 invalid device error` then we could link them to the pull requests with fixes.

**Comment by sveinse** [comment]
> @sveinse I am glad to hear we got some light in the end of tunnel. Do I understand it correctly that proposed modification starting from line 5992 improved situation with glitches?

Affirmative. That fix resolves any startup glitches.

**Comment by sveinse** [comment]
@dmitrykos Since the initial buffer fill fix was found last, I wanted to verify if the /2 fix was necessary. I did this by commenting out https://github.com/PortAudio/portaudio/blob/aecafb1e7d74b2d1705c3a70eea11e3b07301d8c/src/hostapi/wasapi/pa_win_wasapi.c#L5841-L5846. The results were negative. This code is needed to fix glitching. Specifically buffer sizes of 1056 and 528 were glitching continuously. All the other sizes had perfect playback.

**Comment by sveinse** [comment]
I've added a special build of my end-user app for windows with the propsals in PR #307 if anyone else wants to test. It can be downloaded from here: https://github.com/sveinse/elns-portaudio-test/releases/tag/2

**Comment by dmitrykos** [comment]
@sveinse thank you very much for testing, I am really glad that finally we nailed the bug! :)

To summarize (as per @philburk's advise) what we have done with #307 to get rid of audible glitches in Polling Shared mode: 

1. WASAPI implementation was not setting system timer function's granularity and therefore `WaitForSingleObject` was constantly expiring because standard granularity is 15 ms, now once WASAPI thread is started it will force system timer granularity to 1 ms in Poll and Even modes with `timeBeginPeriod`. Granularity is restored back to its default value when thread exits (stream is stopped).

2. If user buffer is 1/2 of the host buffer we will wait 2 times less than the duration of the user buffer as it helped to get rid of glitching in our tests.

3. To avoid underrun due to sleeping again if polling cycle resulted in less frames than we expected while sleeping, we will re-try polling without sleeping (0 ms). Such short busy waiting results in 1-2 additional iterations that is ok and will not harm the performance and power saving.

4. To avoid occasional glitches when stream is started and requested latency is low (22 ms) we have to preload whole host buffer, not just with chunk of the user buffer size.

I hope I did not miss anything else. Will proceed with merging into the master.
